pkg/vault/contracts/VaultAdmin.sol

```diff
@@ -438,41 +442,92 @@ contract VaultAdmin is IVaultAdmin, VaultCommon, Authentication {
             revert InvalidUnderlyingTokenAsset();
         }
 
+        // Register asset of wrapper, so it cannot change.
+        _bufferAssets[wrappedToken] = underlyingToken;
+
+        // Take debt for initialization assets.
+        _takeDebt(IERC20(underlyingToken), amountUnderlyingRaw);
+        _takeDebt(wrappedToken, amountWrappedRaw);
+
+        // Update buffer balances.
+        bytes32 bufferBalances = _bufferTokenBalances[wrappedToken];
+        bufferBalances = PackedTokenBalance.toPackedBalance(
+            bufferBalances.getBalanceRaw() + amountUnderlyingRaw,
+            bufferBalances.getBalanceDerived() + amountWrappedRaw
+        );
+
+        _bufferTokenBalances[wrappedToken] = bufferBalances;
+
         // Amount of shares to issue is the total underlying token that the user is depositing.
         issuedShares = wrappedToken.convertToAssets(amountWrappedRaw) + amountUnderlyingRaw;
 
-        if (_bufferAssets[wrappedToken] == address(0)) {
-            // Buffer is not initialized yet, so we initialize it.
+        _ensureMinimumTotalSupply(issuedShares);
 
-            // Register asset of wrapper, so it cannot change.
-            _bufferAssets[wrappedToken] = underlyingToken;
+        issuedShares -= _MINIMUM_TOTAL_SUPPLY;
+        _mintBufferShares(wrappedToken, sharesOwner, issuedShares);
+        _mintMinimumBufferSupplyReserve(wrappedToken);
 
-            // Burn MINIMUM_TOTAL_SUPPLY shares, so the buffer can never go back to zero liquidity
-            // (avoids rounding issues with low liquidity).
-            _bufferTotalShares[wrappedToken] = _MINIMUM_TOTAL_SUPPLY;
-            issuedShares -= _MINIMUM_TOTAL_SUPPLY;
-        } else if (_bufferAssets[wrappedToken] != underlyingToken) {
-            // Asset was changed since the first bufferAddLiquidity call.
-            revert WrongWrappedTokenAsset(address(wrappedToken));
-        }
+        emit LiquidityAddedToBuffer(wrappedToken, amountUnderlyingRaw, amountWrappedRaw);
+    }
 
-        bytes32 bufferBalances = _bufferTokenBalances[wrappedToken];
+    /// @inheritdoc IVaultAdmin
+    function addLiquidityToBuffer(
+        IERC4626 wrappedToken,
+        uint256 amountUnderlyingRaw,
+        uint256 amountWrappedRaw,
+        address sharesOwner
+    )
+        public
+        onlyVaultDelegateCall
+        onlyWhenUnlocked
+        whenVaultBuffersAreNotPaused
+        withInitializedBuffer(wrappedToken)
+        nonReentrant
+        returns (uint256 issuedShares)
+    {
+        // Check wrapped token asset correctness.
+        address underlyingToken = wrappedToken.asset();
+        _ensureCorrectBufferAsset(wrappedToken, underlyingToken);
```

Comment:
Maybe these lines could be inside the "withInitializedBuffer" modifier, since they'll repeat in each function that requires an initialized buffer.

Reply:
I tried all that (see my comment above). Everything increases the bytecode size.

Reply:
the problem is that we need the underlying token besides the check. I'd evaluate other options after the first pass; I think this part is clear enough like this for now. It's also the most straightforward way considering the starting point anyways.

Reply:
That's not exactly the case.

`remove` needs initialized buffer (asset != 0), but shouldn't need to check whether the current wrapper asset matches. That way it works more similarly to a remove liquidity recovery, where no external calls are made.

For good wrappers it should always be the case, and checking whether the asset changed in the same step as checking whether the asset is not address(0) would make sense too.